Parse MaterialX metadata into SdrProperty

seando-adsk opened the pull request:
### Description of Change(s)
Fixes #1874 

- "enum" and "enumvalues" are parsed as options
- "doc" is parsed as help
- "uiname" is parsed as label
- "uifolder" is parsed as page
- all other MaterialX metadata is added as hints

<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [x] I have submitted a signed Contributor License Agreement

JGamache-autodesk committed:
Parse MaterialX metadata into SdrProperty

Fixes #1874

- "enum" and "enumvalues" are parsed as options
- "doc" is parsed as help
- "uiname" is parsed as label
- "uifolder" is parsed as page
- all other MaterialX metadata is added as hints

JGamache-autodesk committed:
Improve readability

JGamache-autodesk committed:
Add unit documentation when possible.

jilliene commented:
Filed as internal issue #USD-7414

JGamache-autodesk committed:
Added defaultgeomprop. Cleaned up strings.

spiffmon commented on pxr/usd/usdMtlx/parser.cpp:
```diff
@@ -207,14 +296,13 @@ ShaderBuilder::AddProperty(
         // always-required primvars; note that this means we might overestimate
         // which primvars are referenced in a material.
-        const auto& defaultgeomprop = element->getAttribute(defaultgeompropName);
+        const auto& defaultgeomprop = element->getAttribute(_tokens->defaultgeomprop.GetString());
```
TfToken should auto-convert to std::string

spiffmon commented on pxr/usd/usdMtlx/parser.cpp:
```diff
@@ -239,6 +327,45 @@ ShaderBuilder::AddProperty(
     }
 
+    if (!isOutput) {
+        ParseMetadata(metadata, SdrPropertyMetadata->Label, element, "uiname");
+        ParseMetadata(metadata, SdrPropertyMetadata->Help, element, "doc");
+        ParseMetadata(metadata, SdrPropertyMetadata->Page, element, "uifolder");
```
Any reason why these three literals, plus "enum" and "enumvalues" above aren't also added to the private tokens?

JGamache-autodesk replied:
Yes, but not necessarily a good one. These strings are only used as std::string on the MaterialX side of things. Would be best to have them as tokens as well since that prevents an unnecessary malloc/strcpy.

spiffmon commented on pxr/usd/usdMtlx/parser.cpp:
```diff
@@ -239,6 +327,45 @@ ShaderBuilder::AddProperty(
+
+        if (!metadata.count(SdrPropertyMetadata->Help) &&
+             metadata.count(_tokens->unit)) {
+            // The unit can be helpful if there is no documentation.
```
Not sure what `unittype` is, but would it also be useful?  Seems like `defaultgeomprop` would also be handy to provide?

JGamache-autodesk replied:
I see only two places in MaterialX where unittype is defined. There are a few angles defined as being "unittype = radian" and one thin_film thickness being tagged with "unittype = nanometer". I suspect ST coordinates could be in meters in some future, but we are not there yet.

spiffmon replied:
Oh hmmm...   these will want to be **very prominently** documented, as they diverge from USD norms.  All angles are degrees in USD schemas, and linear units are generally relative to stage `metersPerUnit` .  So yeah, I'd double-down on wanting to fold unittype into the description, if the nodes that use them don't already provide documentation that mentions them!

spiffmon commented:
Just a couple questions/sugs, but looks good overall, @JGamache-autodesk .  Let us know if you plan on updating the PR?

spiffmon commented:
@JGamache-autodesk , we might be able to get this into 23.02 if you can update the PR this week/early next week?

JGamache-autodesk committed:
Cleanup use of C strings

All C strings were replaced by TfTokens.
Reinforced warning about unit divergence between MaterialX and USD.

JGamache-autodesk commented:
Hi @spiffmon, I have a patch in my local repo. I will try to get it merged into the PR early next week.

seando-adsk committed:
Merge branch 'dev' into t_gamaj/USD-28/parse_materialx_metadata

spiffmon commented on pxr/usd/usdMtlx/parser.cpp:
```diff
@@ -48,6 +48,26 @@ TF_DEFINE_PRIVATE_TOKENS(
+    (defaultgeomprop)
+    (defaultinput)
+    (doc)
+    (enum)
```
@JGamache-autodesk , did this compile for you?  `enum` is not a legal token/identifier name in C++.  YOu can do something like:
```
((enum_, "enum"))
```

Thanks!

seando-adsk replied:
@spiffmon I added the patch (from Jerry).

spiffmon replied:
Looks great - thanks, @seando-adsk !

JGamache-autodesk committed:
Fix invalid use of C++ identifier
